src/java.base/share/classes/jdk/internal/foreign/abi/fallback/FFIType.java

```diff
@@ -0,0 +1,145 @@
+ * } ffi_type;
+ */
+class FFIType {
+    private static final ValueLayout SIZE_T = switch ((int) ADDRESS.bitSize()) {
+            case 64 -> JAVA_LONG;
+            case 32 -> JAVA_INT;
+            default -> throw new IllegalStateException("Address size not supported: " + ADDRESS.byteSize());
+        };
+    private static final ValueLayout UNSIGNED_SHORT = JAVA_SHORT;
+    private static final StructLayout LAYOUT = Utils.computePaddedStructLayout(
+            SIZE_T, UNSIGNED_SHORT, UNSIGNED_SHORT.withName("type"), ADDRESS.withName("elements"));
+
+    private static final VarHandle VH_TYPE = LAYOUT.varHandle(MemoryLayout.PathElement.groupElement("type"));
+    private static final VarHandle VH_ELEMENTS = LAYOUT.varHandle(MemoryLayout.PathElement.groupElement("elements"));
+    private static final VarHandle VH_SIZE_T_ARRAY = SIZE_T.arrayElementVarHandle();
+
+    private static MemorySegment make(List<MemoryLayout> elements, FFIABI abi, Arena scope) {
+        MemorySegment elementsSeg = scope.allocate((elements.size() + 1) * ADDRESS.byteSize());
+        int i = 0;
+        for (; i < elements.size(); i++) {
+            MemoryLayout elementLayout = elements.get(i);
+            MemorySegment elementType = toFFIType(elementLayout, abi, scope);
+            elementsSeg.setAtIndex(ADDRESS, i, elementType);
+        }
+        // elements array is null-terminated
+        elementsSeg.setAtIndex(ADDRESS, i, MemorySegment.NULL);
+
+        MemorySegment ffiType = scope.allocate(LAYOUT);
+        VH_TYPE.set(ffiType, LibFallback.STRUCT_TAG);
+        VH_ELEMENTS.set(ffiType, elementsSeg);
+
+        return ffiType;
+    }
+
+    private static final Map<Class<?>, MemorySegment> CARRIER_TO_TYPE = Map.of(
+        boolean.class, LibFallback.UINT8_TYPE,
+        byte.class, LibFallback.SINT8_TYPE,
+        short.class, LibFallback.SINT16_TYPE,
+        char.class, LibFallback.UINT16_TYPE,
+        int.class, LibFallback.SINT32_TYPE,
+        long.class, LibFallback.SINT64_TYPE,
+        float.class, LibFallback.FLOAT_TYPE,
+        double.class, LibFallback.DOUBLE_TYPE,
+        MemorySegment.class, LibFallback.POINTER_TYPE
+    );
+
+    static MemorySegment toFFIType(MemoryLayout layout, FFIABI abi, Arena scope) {
+        if (layout instanceof GroupLayout grpl) {
+            if (grpl instanceof StructLayout strl) {
+                // libffi doesn't want our padding
+                List<MemoryLayout> filteredLayouts = strl.memberLayouts().stream()
+                        .filter(Predicate.not(PaddingLayout.class::isInstance))
+                        .toList();
+                MemorySegment structType = make(filteredLayouts, abi, scope);
+                verifyStructType(strl, filteredLayouts, structType, abi);
+                return structType;
+            }
+            assert grpl instanceof UnionLayout;
+            throw new IllegalArgumentException("Fallback linker does not support by-value unions: " + grpl);
```

Comment:
For posterity, I suggest adding a comment here with the the bug number.
```suggestion
            // JDK-8301800
            throw new IllegalArgumentException("Fallback linker does not support by-value unions: " + grpl);
```